content/asm_3.md

```diff
@@ -11,243 +14,604 @@ global _start
 
-* `push argument` - increments stack pointer (RSP) and stores argument in location pointed by stack pointer
-* `pop argument` - copied data to argument from location pointed by stack pointer
+![asm-3-stack-fram-of__double-2](./assets/asm-3-stack-of__double-2.svg)
 
-Let's look on one simple example:
+Finally, we put the value from the stack at offset `-20` (the value of the function's parameter) into the `eax` register and multiply it by `2`, which is located on the stack at offset `-4`. The result of the multiplication is then stored in the `eax` register. This simple example shows how the stack is used to access both parameters and local variables of a function.
 
-```assembly
-global _start
+## Stack operations
 
-section .text
+We already have seen two assembly instructions that affects the current state of the stack:
 
-_start:
-		mov rax, 1
-		mov rdx, 2
-		push rax
-		push rdx
+- `push` - pushes the operand into the stack.
+- `pop` - pops the top value from the stack.
 
-		mov rax, [rsp + 8]
+x86_64 processors provide additional instruction that brings affect on the stack. Besides those instruction we also have seen familiar to us:
 
-		;;
-		;; Do something
-		;;
+- `call`
+- `ret`
+
+The first one instruction calls the given procedure. It affects stack by saving the return address on the stack before call. The second instruction is an "exit" from the given procedure. It affects the stack by removing the return address from the stack and transferring the execution flow to it. 
+
+In the [previous post](asm_2.md) we got familiar with the with such a concepts as [function prologue and epilogue](https://en.wikipedia.org/wiki/Function_prologue_and_epilogue). These are special instructions that we usually can meet in the beginning and in the end of the function:
+
+```assembly
+foo:
+        ;; Function prologue
+        push %rbp
+        mov     rbp, rsp
+
+        ;;
+        ;; Function body
+        ;;
+
+        ;; Function epilogue
+        mov rsp, rbp
+        pop
 ```
-Here we can see that we put 1 to rax register and 2 to rdx register. After it we push to stack values of these registers. Stack works as LIFO (Last In First Out). So after this stack or our application will have following structure:
 
-![stack diagram](/content/assets/stack-diagram.png)
+These two could be replaced with special instructions: `enter N, 0` and `leave`. The first instruction has two operands: 
 
-Then we copy value from stack which has address rsp + 8. It means we get address of top of stack, add 8 to it and copy data by this address to rax. After it rax value will be 1.
+- Number of bytes that needs to be subtracted from the `rsp` register to allocate space on stack.
+- Number of levels of stack frames in nested calls.
 
-## Example
+These both instructions are considered "outdated" but still will work because of backward compatibility.
```

Comment:
```suggestion
These instructions are considered "outdated" but still work because of backward compatibility.
```

Reply:
Why are they outdated? By what are they replaced?

Reply:
Good note, added the reason